larq_zoo/train.py

```diff
@@ -1,65 +1,132 @@
-        verbose=2 if tensorboard else 1,
-        initial_epoch=initial_epoch,
-        callbacks=callbacks,
-    )
 
-    model_name = build_model.__name__
-    model.save_weights(path.join(output_dir, f"{model_name}_weights.h5"))
+    @property
+    def model_path(self):
+        return self.output_dir / "model"
 
-    # Save weights without top
-    notop_model = build_model(
-        hparams, **dataset.preprocessing.kwargs, include_top=False
+    metrics: List[Union[Callable[[tf.Tensor, tf.Tensor], float], str]] = Field(
+        lambda: ["sparse_categorical_accuracy", "sparse_top_k_categorical_accuracy"]
     )
-    notop_model.set_weights(model.get_weights()[: len(notop_model.get_weights())])
-    notop_model.save_weights(path.join(output_dir, f"{model_name}_weights_notop.h5"))
+
+    loss = Field("sparse_categorical_crossentropy")
+
+    @Field
+    def callbacks(self) -> List[tf.keras.callbacks.Callback]:
+        callbacks = []
+        if self.use_model_checkpointing:
+            callbacks.append(
+                utils.ModelCheckpoint(
+                    filepath=str(self.model_path), save_weights_only=True
+                )
+            )
+        if hasattr(self, "learning_rate_schedule"):
+            callbacks.append(
+                keras.callbacks.LearningRateScheduler(self.learning_rate_schedule)
+            )
+        if self.use_tensorboard:
+            callbacks.append(
+                keras.callbacks.TensorBoard(
+                    log_dir=self.output_dir, write_graph=False, profile_batch=0
+                )
+            )
+        return callbacks
+
+    def run(self):
+        self.output_dir.mkdir(parents=True, exist_ok=True)
+
+        initial_epoch = utils.get_current_epoch(self.output_dir)
+
+        train_data, num_train_examples = self.dataset.train(
+            decoders=self.preprocessing.decoders
+        )
+        train_data = (
+            train_data.cache()
+            .shuffle(10 * self.batch_size)
+            .repeat()
+            .map(
+                functools.partial(self.preprocessing, training=True),
+                num_parallel_calls=tf.data.experimental.AUTOTUNE,
+            )
+            .batch(self.batch_size)
+            .prefetch(1)
+        )
+
+        validation_data, num_validation_examples = self.dataset.validation(
+            decoders=self.preprocessing.decoders
+        )
+        validation_data = (
+            validation_data.cache()
+            .repeat()
+            .map(self.preprocessing, num_parallel_calls=tf.data.experimental.AUTOTUNE)
+            .batch(self.batch_size)
+            .prefetch(1)
+        )
+
+        with utils.get_distribution_scope(self.batch_size):
+            self.model.compile(
+                optimizer=self.optimizer, loss=self.loss, metrics=self.metrics,
+            )
+
+            lq.models.summary(self.model)
+
+            if initial_epoch > 0:
+                self.model.load_weights(self.model_path)
```

Comment:
Does this correctly restore the optimizer state? I see we don't use a checkpoint manager here

Reply:
In general, this class contains a lot of duplicate code with the internal `PolyaxonExperiment` we have. Does it make sense to add some abstraction of this to zookeeper instead?

Reply:
Talked to @koenhelwegen about this; the training loop is much less important than the models, and we can push any changes that aren't bug fixes to future PRs.

Reply:
I believe that you might be right about this not restoring optimizer state (at least not for multi-GPU) but as this was the previous implementation (https://github.com/larq/zoo/blob/2e62a7a0dc266cbaf4d56a1d1e95703b54eeb06a/larq_zoo/train.py#L39-L41) I think we can leave it for now. Fixing that can be pushed to a future PR.

Reply:
definitely wouldn't worry about multi-gpu training or restoring. The training loops are here so people can see what we have done, meaning simplicity should be prioritized over supporting a range of complex infrastructure options.